src/eytzinger_benchmark.zig

```diff
@@ -0,0 +1,319 @@
+                .type = PERF.TYPE.HARDWARE,
+                .config = @enumToInt(counter),
+                .flags = .{
+                    .disabled = true,
+                    .exclude_kernel = true,
+                    .exclude_hv = true,
+                },
+            };
+            perf_fds[i] = try perf_event_open(&attr, 0, -1, perf_fds[0], PERF.FLAG.FD_CLOEXEC);
+        }
+        const err = std.os.linux.ioctl(perf_fds[0], PERF.EVENT_IOC.ENABLE, PERF.IOC_FLAG_GROUP);
+        if (err == -1) return error.Unexpected;
+
+        // Start the wall clock after perf, since setup is slow.
+        const timer = try std.time.Timer.start();
+        return Benchmark{
+            .timer = timer,
+            // TODO pass std.os.linux.rusage.SELF once Zig is upgraded
+            .rusage = std.os.getrusage(0),
+            .perf_fds = perf_fds,
+        };
+    }
+
+    fn end(self: *Benchmark, samples: usize) !BenchmarkResult {
+        defer {
+            for (perf_counters) |_, i| {
+                std.os.close(self.perf_fds[i]);
+                self.perf_fds[i] = -1;
+            }
+        }
+        const rusage = std.os.getrusage(0);
+        const err = std.os.linux.ioctl(self.perf_fds[0], PERF.EVENT_IOC.DISABLE, PERF.IOC_FLAG_GROUP);
+        if (err == -1) return error.Unexpected;
+        return BenchmarkResult{
+            .wall_time = self.timer.read() / samples,
+            .utime = (timeval_to_ns(rusage.utime) - timeval_to_ns(self.rusage.utime)) / samples,
+            .cpu_cycles = (try readPerfFd(self.perf_fds[0])) / samples,
+            .instructions = (try readPerfFd(self.perf_fds[1])) / samples,
+            .cache_references = (try readPerfFd(self.perf_fds[2])) / samples,
+            .cache_misses = (try readPerfFd(self.perf_fds[3])) / samples,
+            .branch_misses = (try readPerfFd(self.perf_fds[4])) / samples,
+        };
+    }
+};
+
+// shuffle([0,1,…,n-1])
+fn shuffled_index(comptime n: usize, rand: *std.rand.Random) [n]usize {
+    var indices: [n]usize = undefined;
+    for (indices) |*i, j| i.* = j;
+    rand.shuffle(usize, indices[0..]);
+    return indices;
+}
+
+fn timeval_to_ns(tv: std.os.timeval) u64 {
+    const ns_per_us = std.time.ns_per_s / std.time.us_per_s;
+    return @bitCast(u64, tv.tv_sec) * std.time.ns_per_s +
+        @bitCast(u64, tv.tv_usec) * ns_per_us;
+}
+
+fn readPerfFd(fd: std.os.fd_t) !usize {
+    var result: usize = 0;
+    const n = try std.os.read(fd, std.mem.asBytes(&result));
+    assert(n == @sizeOf(usize));
+    return result;
+}
+
+fn binary_search_keys(
+    comptime layout: Layout,
+    comptime Key: type,
+    comptime Val: type,
+    comptime compare_keys: fn (Key, Key) math.Order,
+    summary: []const Key,
+    values: []const Val,
+    key: Key,
+) []const Val {
+    assert(summary.len == layout.keys_count);
+    assert(values.len == layout.values_count);
+    const key_index = binary_search(Key, Key, Val.key_from_key, compare_keys, summary, key);
```

Comment:
`Val.key_from_key` should be called `key_from_value` here, which is how it is named in `binary_search.zig`

Reply:
`Val.key_from_value` is a distinct function. `key_from_key` mimics its interface, but it doesn't actually involve the "value" (in this case, the value is just a key, so it's just the identity function). The function is only attached to the `Val` struct because that is where the `Key` type is defined.